### PR TITLE
[SPARK-4233] [SPARK-4367] [SPARK-3947] [SPARK-3056] [SQL] Aggregation Improvement

### PR DESCRIPTION
**Pull request by yhuai**
This is the first PR for the aggregation improvement, which is tracked by https://issues.apache.org/jira/browse/SPARK-4366 (umbrella JIRA). This PR contains work for its subtasks, SPARK-3056, SPARK-3947, SPARK-4233, and SPARK-4367.

This PR introduces a new code path for evaluating aggregate functions. This code path is guarded by `spark.sql.useAggregate2` and by default the value of this flag is true.

This new code path contains:
- A new aggregate function interface (`AggregateFunction2`) and 7 built-int aggregate functions based on this new interface (`AVG`, `COUNT`, `FIRST`, `LAST`, `MAX`, `MIN`, `SUM`)
- A UDAF interface (`UserDefinedAggregateFunction`) based on the new code path and two example UDAFs (`MyDoubleAvg` and `MyDoubleSum`).
- A sort-based aggregate operator (`Aggregate2Sort`) for the new aggregate function interface .
- A sort-based aggregate operator (`FinalAndCompleteAggregate2Sort`) for distinct aggregations (for distinct aggregations the query plan will use `Aggregate2Sort` and `FinalAndCompleteAggregate2Sort` together).

With this change, `spark.sql.useAggregate2` is `true`, the flow of compiling an aggregation query is:
1. Our analyzer looks up functions and returns aggregate functions built based on the old aggregate function interface.
2. When our planner is compiling the physical plan, it tries try to convert all aggregate functions to the ones built based on the new interface. The planner will fallback to the old code path if any of the following two conditions is true:
- code-gen is disabled.
- there is any function that cannot be converted (right now, Hive UDAFs).
- the schema of grouping expressions contain any complex data type.
- There are multiple distinct columns.

Right now, the new code path handles a single distinct column in the query (you can have multiple aggregate functions using that distinct column). For a query having a aggregate function with DISTINCT and regular aggregate functions, the generated plan will do partial aggregations for those regular aggregate function.

Thanks @chenghao-intel for his initial work on it.

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/FunctionRegistry.scala [hunk 1]
```diff
@@ -146,6 +147,7 @@ object FunctionRegistry {
     expression[Max]("max"),
     expression[Min]("min"),
     expression[Sum]("sum"),
+    expression[MyDoubleSum]("mydoublesum"),
```
Comment (yhuai) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/FunctionRegistry.scala]: `MyDoubleSum` is just an example. It will be removed later.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala [hunk 2]
```diff
@@ -537,6 +541,7 @@ class Analyzer(
     def containsAggregates(exprs: Seq[Expression]): Boolean = {
       exprs.foreach(_.foreach {
         case agg: AggregateExpression => return true
```
Comment (rxin) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: if we are not going to be removing the old aggregate expression for a while, how about creating a new aggregateexpression that's the shared parent, and rename existing one aggregateexpression1? i worry we miss places that we match on.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Expression.scala [hunk 3]
```diff
@@ -90,7 +90,7 @@ abstract class Expression extends TreeNode[Expression] {
     val primitive = ctx.freshName("primitive")
     val ve = GeneratedExpressionCode("", isNull, primitive)
     ve.code = genCode(ctx, ve)
-    ve
```
Comment (rxin) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Expression.scala]: for very large expressions, this can be pretty bad. i wonder if we should have a debug mode.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate2/Aggregate2Sort.scala [hunk 4]
```diff
@@ -0,0 +1,365 @@
...
+
+  override def requiredChildOrdering: Seq[Seq[SortOrder]] =
+    groupingExpressions.map(SortOrder(_, Ascending)) :: Nil
+
+  override def output: Seq[Attribute] = resultExpressions.map(_.toAttribute)
+
+  protected override def doExecute(): RDD[InternalRow] = attachTree(this, "execute") {
+    child.execute().mapPartitions { iter =>
+
+      new Iterator[InternalRow] {
```
Comment (rxin) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate2/Aggregate2Sort.scala]: we should take this iterator out as a named iterator

File: sql/hive/src/test/java/test/org/apache/spark/sql/hive/aggregate2/MyJavaUDAF.java [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+  }
+
+  @Override public Object evaluate(Row buffer) {
+    if (buffer.isNullAt(0)) {
+      return null;
+    } else {
+      return buffer.getDouble(0);
+    }
+  }
+}
```
Comment (yhuai) [inline sql/hive/src/test/java/test/org/apache/spark/sql/hive/aggregate2/MyJavaUDAF.java]: @marmbrus @rxin @davies This is an example UDAF in JAVA.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/expressions/aggregate2/udaf.scala [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+
+import org.apache.spark.Logging
+import org.apache.spark.sql.catalyst.expressions.codegen.GenerateMutableProjection
+import org.apache.spark.sql.catalyst.{InternalRow, CatalystTypeConverters}
+import org.apache.spark.sql.catalyst.expressions._
+import org.apache.spark.sql.catalyst.expressions.aggregate2.{InputAggregationBuffer, AggregateFunction2, MutableAggregationBuffer}
+import org.apache.spark.sql.types._
+import org.apache.spark.sql.Row
+
+abstract class UserDefinedAggregateFunction extends Serializable {
```
Comment (yhuai) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/expressions/aggregate2/udaf.scala]: @marmbrus @rxin @davies This is the UDAF interface.

**Comment by yhuai** [comment]
Just added an UDAF interface. Will add comments and doc tomorrow.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate2/Aggregate2Sort.scala [hunk 7]
```diff
@@ -0,0 +1,365 @@
...
+case class Aggregate2Sort(
+    groupingExpressions: Seq[NamedExpression],
+    aggregateExpressions: Seq[AggregateExpression2],
+    aggregateAttributes: Seq[Attribute],
+    resultExpressions: Seq[NamedExpression],
+    child: SparkPlan)
+  extends UnaryNode {
+
+  /** Indicates if this operator is for partial aggregations. */
+  val partialAggregation: Boolean = {
```
Comment (chenghao-intel) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate2/Aggregate2Sort.scala]: Is there any consideration that we don't use the `partialAggregation` passed from `SparkStrategies`?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate2/Aggregate2Sort.scala (hunk 4 above)
Comment (chenghao-intel) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate2/Aggregate2Sort.scala]: Probably we'd better split the partial / final aggregation Iterator into different class. As the partial aggregation(pre-shuffle) can be followed by hash/sort aggregation(post-shuffle) for the further improvement.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate2/aggregates.scala [hunk 8]
```diff
@@ -0,0 +1,279 @@
...
+
+  self: Product =>
+
+  /** An aggregate function is not foldable. */
+  override def foldable: Boolean = false
+
+  /**
+   * The offset of this function's buffer in the underlying buffer shared with other functions.
+   */
+  var bufferOffset: Int = 0
```
Comment (chenghao-intel) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate2/aggregates.scala]: Instead of explicit expose the `bufferOffset`, probably people prefer using the `BoundReference`, it's created by `BindReferences.bindReference`, I think we can pass in the associated `BoundReference` (bind with bufferSchema) while `initialize` the `AggregateFunction2` in each executor.

We can add the implicit function in this class, to convert `BoundReference` => Int. So people can write code like:
`mutableRow.isNullAt(BoundReference)`

File: sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate2/Aggregate2Sort.scala (hunk 7 above)
Comment (yhuai) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate2/Aggregate2Sort.scala]: The planner does not need to pass in `partialAggregation`. The operator will figure it out based on modes of those functions.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate2/aggregates.scala (hunk 8 above)
Comment (yhuai) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate2/aggregates.scala]: Thanks for the suggestion.

`bufferOffset` is an internal thing. For our built-in aggregate functions, we will try to implement them based on `AlgebraicAggregate`, which supports code-generation. The developers of `AlgebraicAggregate` do not need to use `bufferOffset`. For UDAFs, they should be implemented based on `UserDefinedAggregateFunction`. We will hide this offset from UDAFs. So, hopefully developers will not need to use `bufferOffset` at all. 

I will think about this part and see if there is any other way. I will also go through those new interfaces to make sure we properly set their access modifiers.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Expression.scala (hunk 3 above)
Comment (marmbrus) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Expression.scala]: I think it is probably worth it.  When codegen bugs pop-up in the real world this makes it possible to debug, where as otherwise the code is super hard to follow.  I tested on a pretty deeply nested expression and the code size only went from 25kb to 50kb.  I'd imagine the parser is pretty good at skipping over comments.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate2/aggregates.scala [hunk 9]
```diff
@@ -0,0 +1,283 @@
...
+
+/**
+ * An [[AggregateFunction2]] with [[Partial]] mode is used to evaluate this function directly
+ * from original input rows without any partial aggregation.
+ * This function updates the given aggregation buffer with the original input of this
+ * function. When it has processed all input rows, the final result of this function is returned.
+ */
+private[sql] case object Complete extends AggregateMode
+
+private[sql] case object NoOp extends Expression {
```
Comment (marmbrus) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate2/aggregates.scala]: We should move this out and comment it.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate2/Aggregate2Sort.scala (hunk 4 above)
Comment (yhuai) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate2/Aggregate2Sort.scala]: @chenghao-intel Yeah, that's a good idea.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate2/aggregates.scala [hunk 10]
```diff
@@ -0,0 +1,237 @@
...
+
+/**
+ * An [[AggregateFunction2]] with [[Partial]] mode is used to evaluate this function directly
+ * from original input rows without any partial aggregation.
+ * This function updates the given aggregation buffer with the original input of this
+ * function. When it has processed all input rows, the final result of this function is returned.
+ */
+private[sql] case object Complete extends AggregateMode
+
+private[sql] case object NoOp extends Expression with Unevaluable {
```
Comment (yhuai) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate2/aggregates.scala]: Will add comments to it and figure out the right place for it later.

**Comment by yhuai** [comment]
Seems https://github.com/apache/spark/commit/b285ac5ba85fe0b32b00726ad7d3a2efb602e885 did not add golden files for `udf_unhex`. Those files are included in this PR.

**Comment by yhuai** [comment]
I will merge this one once it passes jenkins to unblock other work. If you have any comments to this, feel free to leave them at here. I will address them in a follow-up PR.

**Comment by yhuai** [comment]
test this please

**Comment by rxin** [comment]
I'm going to merge this epic change!

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/functions.scala [hunk 11]
```diff
@@ -0,0 +1,292 @@
...
+}
+
+case class First(child: Expression) extends AlgebraicAggregate {
+
+  override def children: Seq[Expression] = child :: Nil
+
+  override def nullable: Boolean = true
+
+  // First is not a deterministic function.
+  override def deterministic: Boolean = false
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/functions.scala]: Why did you mark `First` nondeterministic? `First` will give same result for same input all the time, right?

Comment (yhuai) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/functions.scala]: First and Last are both non-deterministic. Let's say you have a query like `SELECT key, first(value) FROM table GROUP BY key`. Since value is not ordered, we do not know what result that `first(value)` will return. I think the only case first and last can generate deterministic results are when you use it as a window function and the argument of it is the order by expression, e.g. `SELECT fist(value) OVER (ORDER BY value) FROM table`.

Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/functions.scala]: OK I'm afraid we may need to add a new concept of mutable expression.

The origin motivation that we add nondeterministic concept is we need some special handling for them:

1. stop doing project collapse for nondeterministic expression.
2. stop pushing down filter through project for nondeterministic expression.
3. pull out nondeterministic expression from sort and materialize them.
4. initialize nondeterministic expression before evaluate them.

AFAIK, we need to do these stuff for `rand`, `randn`, `monotonicallyIncreasingId`, which have mutable states inside. Logically `First` is nondeterministic, but we don't need handle it specially, right? Currently we replace old aggregation expression with new ones during physical plan, so all these special handling won't apply to new aggregation expressions, which is not safe in the future.

cc @rxin @marmbrus

Comment (marmbrus) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/functions.scala]: Yeah, I think we are abusing the concept of `deterministic` here (though @yhuai you are correct that given the same data it will not always return the same result).  Regardless, we should probably leave that flag for 1-1 functions that are non-deterministic since Aggregates state management is already special cased.

Comment (yhuai) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/functions.scala]: @cloud-fan An `Aggregate` operator handles the project list by itself. There is no Project on top of an `Aggregate` to generate results. So, the first two rules you mentioned will not fire.

My question is why we do not need to handle non-deterministic aggregation function specially and why should not use `deterministic` for an aggregate function? Let's say we have a query `sqlContext.sql("select i, f, f, f from (select i, first(j) as f, rand() from t group by i, rand()) tmp")` and let's also assume project collapse works with `Aggregate`. It still makes sense to evaluate first once instead of three times, right? It is probably true that `first` will give you the same answer even if you call it three times, but what about other non-deterministic aggregate functions? Regarding predicate pushdown, we cannot pushdown predicates using the result of an aggregate function through an aggregate operator, right?

Comment (yhuai) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/functions.scala]: @cloud-fan We can add comments to explain that `deterministic` of an aggregate function is different from that of a regular 1-1 function like rand. Or, we can leave deterministic as true for aggregate functions. I prefer the first option.
